python/lib/client/dmod/client/request_clients.py

```diff
@@ -1,600 +1,823 @@
-            for page in range(1, (self.last_response.total_pages + 1)):
-                request = DatasetManagementMessage(action=ManagementAction.DOWNLOAD_DATA, dataset_name=dataset_name,
-                                                   data_location=item_name, page=page)
-                self.last_response: DatasetManagementResponse = await self.async_make_request(request)
-                file.write(self.last_response.file_data)
+            # Do initial request outside of generator
+            await self._transport_client.async_send(data=str(request))
+            async for received_data_msg in self._transfer_receiver():
+                file.write(received_data_msg.data)
+            final_data = await self._transport_client.async_recv()
 
-    async def list_datasets(self, category: Optional[DataCategory] = None) -> List[str]:
-        action = ManagementAction.LIST_ALL if category is None else ManagementAction.SEARCH
-        request = DatasetManagementMessage(action=action, category=category)
-        self.last_response = await self.async_make_request(request)
-        return self._parse_list_of_dataset_names_from_response(self.last_response)
+        try:
+            final_response_json = json.loads(final_data)
+        except Exception as e:
+            msg = f"{self.__class__.__name__} failed with {e.__class__.__name__} parsing `{final_data}` to JSON)"
+            return final_response_type(success=False, reason=f"JSON Parse Failure On Final Response", message=msg)
+
+        final_response = final_response_type.factory_init_from_deserialized_json(final_response_json)
+        if final_response is None:
+            return final_response_type(success=False, reason="Failed to Deserialize Final Response")
+        else:
+            return final_response
+
+    async def upload_dataset_item(self, dataset_name: str, item_name: str, source: Path) -> DatasetManagementResponse:
+        if not source.is_file():
+            return DatasetManagementResponse(success=False, reason="Dataset Upload File Not Found",
+                                             message=f"File {source!s} does not exist")
+        try:
+            request, final_response_type = self._request_prep(dataset_name=dataset_name, item_name=item_name,
+                                                              action=ManagementAction.ADD_DATA)
+        # TODO: (later) implement and use DmodAuthenticationFailure
+        except DmodRuntimeError as e:
+            reason = f'{self.__class__.__name__} Upload Request Auth Failure'
+            return MaaSDatasetManagementResponse(success=False, reason=reason, message=str(e))
 
-    async def upload_to_dataset(self, dataset_name: str, paths: List[Path]) -> bool:
+        chunk_size = 1024
+
+        with source.open('r') as file:
+            last_send = False
+            raw_chunk = file.read(chunk_size)
+
+            while True:
+                await self._transport_client.async_send(data=str(message))
```

Comment:
I didnt see a `message` variable in scope. I think you meant, right?

```suggestion
                await self._transport_client.async_send(data=raw_chunk)
```

Reply:
Ahh, I see I think the `message` variable needs to be moved up, right?

```suggestion
            message = DataTransmitMessage(data=raw_chunk, is_last=last_send)
            while True:
                await self._transport_client.async_send(data=str(message))
```

Reply:
Good catch, although your solution isn't quite right.  The initial _DatasetManagementMessage_ is created by the earlier call to `_request_prep`.  Because we need to look ahead in the data in order to figure out whether we are doing our "last_send," both the initial request and the initial "current" chunk of data are prepared outside the loop, and the subsequent message is prepared at the end of the loop.

I'll fix momentarily.

Reply:
Just to confirm, because the variable in question is initialized prior to these specific lines, this comment doesn't get attributed as _Outdated_, but I have applied a fix.

Reply:
Awesome, yeah you fixed what I was talking about. Thanks!